KEP-2915: add proposal for replacing the kubelet-config-x.y usage

neolit123 opened the pull request:
- One-line PR description:
Initial KEP draft.

<!-- link to the k/enhancements issue -->
- Issue link:
https://github.com/kubernetes/enhancements/issues/2915

<!-- other comments or additional information -->
- Other comments:
NONE

neolit123 commented:
hello, this is a proposal that i think we should get in 1.23 and slowly start the deprecation of the old kubelet-config naming (more details in the KEP) and https://github.com/kubernetes/kubeadm/issues/1582

please have a look as the deadline for KEPs is next week (9th of Sept).

/assign @fabriziopandini @SataQiu @pacoxu

vincepri commented:
/assign

randomvariable commented:
/assign

vincepri commented on keps/sig-cluster-lifecycle/kubeadm/2915-kubeadm-replace-kubelet-config-x.y/README.md:
```diff
@@ -0,0 +1,606 @@
+-->
+
+The proposal is to declare the default ConfigMap and RBAC object format of `kubelet-config-x.y`
+deprecated but support it for a GA deprecation period of 1 year (3 kubeadm releases). In parallel,
+introduce objects named `kubelet-config` (without `x.y`) that will become the source of truth.
```
```suggestion
deprecated but support it for a GA deprecation period of 1 year (3 kubeadm releases). In parallel,
introduce an object named `kubelet-config` (without `x.y`) that will become the source of truth.
```

neolit123 replied:
there are multiple objects - ConfigMaps and RBAC rules.

vincepri commented on keps/sig-cluster-lifecycle/kubeadm/2915-kubeadm-replace-kubelet-config-x.y/README.md:
```diff
@@ -0,0 +1,606 @@
+deprecated but support it for a GA deprecation period of 1 year (3 kubeadm releases). In parallel,
+introduce objects named `kubelet-config` (without `x.y`) that will become the source of truth.
+
+The switch will be controlled by a new kubeadm specific feature gate (not a core one) called
+`UnversionedKubeletConfigMap`. Using a feature gate for similar changes is common in changes
```
```suggestion
The switch will be controlled by a new kubeadm specific feature gate (not a core one) in ClusterConfiguration called
`UnversionedKubeletConfigMap`. Using a feature gate for similar changes is common in changes
```

vincepri commented on keps/sig-cluster-lifecycle/kubeadm/2915-kubeadm-replace-kubelet-config-x.y/README.md:
```diff
@@ -0,0 +1,606 @@
+During `upgrade` a new `kubelet-config-x.y+1` ConfigMap and RBAC rules are written where `y+1` is the target
+MINOR upgrade version. The old `kubelet-config-x.y` are not cleaned up.
+
+Encoding the version in the API object names is redundant and creates unnecessary complications for both
+kubeadm maintainers and users that wish to modify the object in the cluster.
```
How would we deal with version skew between the kubelet-config contents and kubelet reading it?

neolit123 replied:
the current implementation of kubeadm does not deal with that. the stored KubeletConfiguration GVK is checked if it is known by a version of kubeadm and if it's unknown and error is reported. so the source of truth is the GVK inside the config and not the version in the name of the ConfigMap.

we expect an old GVK in the ConfigMap to still work with newer kubelets for a few releases, but otherwise the current state is that users have to manually upconvert when the time comes and until the kubelet has something like `kubelet --convert foo.yaml` in which case kubeadm could do it for them.

neolit123 replied:
added some notes about that under "risk 3".

vincepri commented on keps/sig-cluster-lifecycle/kubeadm/2915-kubeadm-replace-kubelet-config-x.y/README.md:
```diff
@@ -0,0 +1,606 @@
+Note that depending on feedback from users the N+x release schedule might be adjusted - e.g.
+extending or shortening the Alpha or Beta periods.
+
+#### First release N (when the deprecation happens)
+
+- The feature gate is added as Alpha and disabled by default.
+- Users can continue using the old naming format or opt-in into the new naming format
+by enabling the feature gate.
```
Instead of having a feature gate, have we considered writing both maps at the same time and switch the functionality over time?

neolit123 replied:
this was my first approach on the topic, but i estimated that it would require a hard switch at some point. i have covered it briefly in Alternatives (could expand more there).

we could start writing both ConfigMaps for a few releases and then stop writing the old format after the deprecation, but it seems there is a problem with reading. kubeadm as a client would have to use the old format for a few releases and then suddenly switch to the new format. that would be a hard switch withing one release without the option for users to opt-out. a feature gate during the Beta period would allow them the opt-out.

for reading, we could implement a soft-switch of sorts (for e.g. one release) where the contents of both ConfigMaps are combined, considering both of them as the source of truth, but that feels very messy - e.g. trying to guess what field value the user really wants to use.

vincepri replied:
> we could start writing both ConfigMaps for a few releases and then stop writing the old format after the deprecation, but it seems there is a problem with reading. kubeadm as a client would have to use the old format for a few releases and then suddenly switch to the new format. that would be a hard switch within one release without the option for users to opt-out. a feature gate during the Beta period would allow them the opt-out.

I agree that a hard switch can be overwhelming for users at a time, although with proper communication and by spreading the change in a 3 or more releases seems like a good way to migrate. I was thinking of something like:

- Write both configmaps in a release (let's say 1.23) + give a release note to users that the older configmaps are going to be non-functional in a future release.
- Start ignoring the older format in 1.24 (making it non-functional) and alerting users through release notes and required changes.
- Removing and cleaning up the older code.

The reason I'm suggesting this approach is to avoid maintaining two code-paths for longer period of times with a feature gate, the solution proposed above seems more a migration than a new feature, although I'm fine either way.

Having a clear cut-date for the migration also helps kubeadm users that have to support multiple versions. Cluster API for example can start detecting the change by version, rather than by version + feature gate and migrating clusters on the fly. I'd assume other users might want to do something similar.

neolit123 replied:
> I agree that a hard switch can be overwhelming for users at a time, although with proper communication and by spreading the change in a 3 or more releases seems like a good way to migrate. I was thinking of something like:

we've decided on a hard switch for the deprecation of the `master` label and taint - without a feature gate over a period of time because a feature gate made it more difficult:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/kubeadm/2067-rename-master-label-taint#using-a-feature-gate
but i still have some concerns that between some of the stages we are going to break users badly.

as the main problem is that a good number of users simply skip reading the release notes.

> The reason I'm suggesting this approach is to avoid maintaining two code-paths for longer period of times with a feature gate, 

the two code paths would be mostly trivial with a ClusterConfiguration.FeatureGates read controlling the formatting of a string.

> the solution proposed above seems more a migration than a new feature, although I'm fine either way.

i'm also fine either way, and would like to hear more opinions before Sept 9th.

> Having a clear cut-date for the migration also helps kubeadm users that have to support multiple versions. Cluster API for example can start detecting the change by version, rather than by version + feature gate and migrating clusters on the fly. I'd assume other users might want to do something similar.

with or without the feature gate idea CAPI would have to adapt for it's upgrade code to populate the missing API objects normally maintained by `kubeadm upgrade apply`. a version check only (==1.24) would be sufficient for CAPI to decide if it has to create a copy of the existing objects into objects with the new name format. i don't think it has to check the state of the feature gate.

AFAIK, CAPI currently has the feature gates of the ClusterConfiguration immutable, which effectively means CAPI users cannot use them:
https://github.com/kubernetes-sigs/cluster-api/issues/4180

if they can use them and if we go the FG route, this means that CAPI should adapt to the "copy logic" above in 1.23 (Alpha, not 1.24 / Beta).

fabriziopandini replied:
I'm +1 for the feature gate given that this allows users some flexibility in deciding when to switch (and given that the implementation effort seems reasonable).
From a CAPI PoV, I will suggest we wait for this change to become beta active but default; this will allows us to just work by version (but with an escape patch via the feature flag if we decide to wait a little bit longer for any reason).

fabriziopandini commented on keps/sig-cluster-lifecycle/kubeadm/2915-kubeadm-replace-kubelet-config-x.y/README.md:
```diff
@@ -0,0 +1,606 @@
+
+NONE
+
+### Risks and Mitigations
```
I wold like a third risk, being kubelet starting to use a new API version for its own configuration, but not providing automatic conversion in place. This will prevent kubeadm to support kubelet N and kubelet N-1 at the same time.

A viable path forward is to use different keys inside the same ConfigMap; in order to make this possible in the future I suggest to start using API version as key in the new ConfigMap instead of kubelet.

neolit123 replied:
i can add this risk in the list.

i think an easier way for us to mitigate would be to ensure that we do not store the latest kubelet API version in the config map for one more release.

for example:
- kubelet API v1beta2 comes out in 1.26.
- for new clusters, kubeadm 1.26 should not write v1beta2 in the CM and should write v1beta1 instead.
- this would allow 1.25 kubelets to join the cluster as well as both 1.26 and 1.25 kubelets would still support v1beta1.

once 1.27 is out (the v1beta2 has been supported for 2 releases):
- new clusters created with kubeadm 1.27 can start writing v1beta2 in the CM.
- 1.26 kubelets would support v1beta2 and they can join the cluster.

i expect sig-node to listen to the requests to support converters on the kubelet CLI (or endpoint) once the v1beta1 is marked as deprecated.

neolit123 replied:
added as "Risk 3", as per my alternative approach above.

fabriziopandini commented on keps/sig-cluster-lifecycle/kubeadm/2915-kubeadm-replace-kubelet-config-x.y/README.md:
```diff
@@ -0,0 +1,606 @@
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Test Plan
```
I think that we need also a using test for spinning up a cluster with the feature gate enabled, but without kubelet skew

neolit123 replied:
do you mean adding an extra kubelet skew test - e.g. kubeadm N deploying kubelet N-1?
we can add such a test when the kubelet v1beta2 is added.

i think i should rename **Skew e2e test job**` in the doc to **Join e2e test job**, because i'm realizing the proposed kubeadm vs control-plane skew test case is not important and the important part is that we ensure "join" gets the correct CM.

neolit123 replied:
> we can add such a test when the kubelet v1beta2 is added.

added under "future e2e tests"

> i think i should rename Skew e2e test job` in the doc to Join e2e test job

done.

fabriziopandini reviewed (commented):
@neolit123 Thanks for driving this effort
Just a couple of nits, but overall lgtm

neolit123 commented:
i have updated the KEP draft again:
- given i didn't see any objections to proceed with the Alpha for 1.23, i pre-marked the KEP as "implementable".
- i implemented my interpretations from @vincepri and @fabriziopandini 's feedback above.

@vincepri @fabriziopandini @SataQiu @randomvariable , PTAL again.
looks like @pacoxu has commented on https://github.com/kubernetes/kubeadm/issues/1582 that he's in favor of the change.

/hold
until review has passed.

neolit123 commented on keps/prod-readiness/sig-cluster-lifecycle/2915.yaml:
```diff
@@ -0,0 +1,3 @@
+kep-number: 2915
+alpha:
+  approver: "@ehashman"
```
/assign @ehashman 

looks like we have to include this file if the KEP is "implementable" for 1.23.

pacoxu commented:
> looks like @pacoxu has commented on [kubernetes/kubeadm#1582](https://github.com/kubernetes/kubeadm/issues/1582) that he's in favor of the change.

The proposal looks good to me. After some testing and code reading about the kubelet-config customization. 
In current codebase, upgrade/init will generate a new version `kubelet-config-x.y` configmap. All `kubeadm join` will use the server's version configmap to set up the kubelet config. The old `kubelet-config-x.y` is ambiguous(For instance, in a 1.22 Kubernetes cluster, when I join a 1.21 node with 1.21 kubeadm, it will take `kubelet-config-1.22` as well which is expected in my mind. Of course, this is not a normal join.).

neolit123 commented:
/assign @johnbelamaric

vincepri commented:
/approve

johnbelamaric commented:
/approve

ehashman commented on keps/prod-readiness/sig-cluster-lifecycle/2915.yaml:
```diff
@@ -0,0 +1,3 @@
+kep-number: 2915
+alpha:
+  approver: "@johnbelamaric"
```
According to the spreadsheet this is me and I don't have a ton of reviews so I'll do it... can you update?

neolit123 replied:
given @johnbelamaric already approved how about we update the spreadsheet for this particular KEP?

ehashman replied:
Do you have a link to that? This is the first I've seen of this KEP, it appears to be net new

RA489 commented:
/lgtm 

Thanks!!

neolit123 commented:
/hold cancel
